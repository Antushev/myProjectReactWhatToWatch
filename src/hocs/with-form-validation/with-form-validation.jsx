import React, {PureComponent} from 'react';

const withFormValidation = (Component) => {
  class WithFormValidation extends PureComponent {
    constructor(props) {
      super(props);

      this.state = {
        inputEmail: ``,
        inputPassword: ``,
        buttonDisabled: true
      };

      this._handleChangeInputEmail = this._handleChangeInputEmail.bind(this);
      this._handleChangeInputPassword = this._handleChangeInputPassword.bind(this);
    }

    render() {
      const {buttonDisabled} = this.state;

      return <Component
        {...this.props}
        buttonDisabled={buttonDisabled}
        handleChangeInputEmail={this._handleChangeInputEmail}
        handleChangeInputPassword={this._handleChangeInputPassword}
      />;
    }

    _handleChangeInputEmail(value) {
      const {inputPassword} = this.state;

      if (inputPassword !== `` && value !== ``) {
        this.setState({
          inputEmail: value,
          buttonDisabled: false
        });
      } else {
        this.setState({
          inputEmail: value,
          buttonDisabled: true
        });
      }
    }

    _handleChangeInputPassword(value) {
      const {inputEmail} = this.state;

      if (inputEmail !== `` && value !== ``) {
        this.setState({
          inputPassword: value,
          buttonDisabled: false
        });
      } else {
        this.setState({
          inputPassword: value,
          buttonDisabled: true
        });
      }
    }
  }

  return WithFormValidation;
};

export {withFormValidation};
