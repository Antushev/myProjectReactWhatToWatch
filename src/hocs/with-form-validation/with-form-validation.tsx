import * as React from 'react';
import {Subtract} from 'utility-types';

interface State {
  inputEmail: string;
  inputPassword: string;
  buttonDisabled: boolean;
}

interface InjectingProps {
  buttonDisabled: boolean;
  onChangeInputEmail: (value: string) => void;
  onChangeInputPassword: (value: string) => void;
}

const withFormValidation = (Component) => {
  type P = React.ComponentProps<typeof Component>;
  type T = Subtract<P, InjectingProps>;

  class WithFormValidation extends React.PureComponent<T, State> {
    constructor(props) {
      super(props);

      this.state = {
        inputEmail: ``,
        inputPassword: ``,
        buttonDisabled: true
      };

      this._handleChangeInputEmail = this._handleChangeInputEmail.bind(this);
      this._handleChangeInputPassword = this._handleChangeInputPassword.bind(this);
    }

    render() {
      const {buttonDisabled} = this.state;

      return <Component
        {...this.props}
        buttonDisabled={buttonDisabled}
        onChangeInputEmail={this._handleChangeInputEmail}
        onChangeInputPassword={this._handleChangeInputPassword}
      />;
    }

    _handleChangeInputEmail(value) {
      const {inputPassword} = this.state;

      if (inputPassword !== `` && value !== ``) {
        this.setState({
          inputEmail: value,
          buttonDisabled: false
        });
      } else {
        this.setState({
          inputEmail: value,
          buttonDisabled: true
        });
      }
    }

    _handleChangeInputPassword(value) {
      const {inputEmail} = this.state;

      if (inputEmail !== `` && value !== ``) {
        this.setState({
          inputPassword: value,
          buttonDisabled: false
        });
      } else {
        this.setState({
          inputPassword: value,
          buttonDisabled: true
        });
      }
    }
  }

  return WithFormValidation;
};

export {withFormValidation};
