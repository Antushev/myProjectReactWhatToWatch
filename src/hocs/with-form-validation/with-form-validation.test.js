import React from 'react';
import renderer from 'react-test-renderer';

import {withFormValidation} from './with-form-validation';

const MockComponent = () => {
  return <div></div>;
};

const MockComponentWrapped = withFormValidation(MockComponent);

describe(`Tests HOC with-form-validation`, () => {
  it(`Test HOC with-form-validation`, () => {
    const tree = renderer
      .create(
          <MockComponentWrapped
            buttonDisabled={true}
            handleChangeInputEmail={() => {}}
            handleChangeInputPassword={() => {}}
          />, {
            createNodeMock() {
              return {};
            }
          }
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
