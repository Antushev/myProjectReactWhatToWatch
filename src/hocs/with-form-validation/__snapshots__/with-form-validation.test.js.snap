// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests HOC with-form-validation Test HOC with-form-validation 1`] = `<div />`;
